[Dashboards as code] Add conversion functions from/to API filters interface and stored filters

nickpeihl opened the pull request:
Fixes #241852 

_Note: Parts of this code were generated or assisted by Copilot. I have reviewed and validated all code for correctness, security, and compliance with Elastic’s standards._

## Summary

Adds bidirectional conversion functions between stored filters (used in saved objects and URL state) and the new AsCodeFilter API format for the Dashboards as Code initiative. This enables seamless migration between the existing filter infrastructure and the new simplified API.

The conversion functions introduced in this PR are not yet used in production. Integrations with Kibana applications like Dashboards, Maps, Lens will be introduced in subsequent PRs.

This PR adds three new packages for As Code Filters. 
* It moves the `AsCodeFiltersSchema` from `@kbn/es-query-server` to a new `@kbn/as-code-filters-schema` package. 
* The transform functions are added to a new `@kbn/as-code-filters-transforms` package. We separate the packages as the schemas are only intended to be used by server code, while the transforms can be used by either server or public code. This separation allows us to avoid bundling large dependencies like `Joi` in client side code.
* The `@kbn/as-code-filters-constants` package allows both the `@kbn/as-code-filters-transforms` and `@kbn/as-code-filters-schema` packages to use the same constants without worrying about circular dependencies and forbidden cross-boundary imports.

## Implementation

**Type-First Routing Approach:**
- Uses `meta.type` from the `FILTERS` enum for deterministic routing to appropriate conversion functions
- Preserves filters without `meta.type`as DSL to prevent data loss

**Conversion Functions:**
- `fromStoredFilter` - Converts StoredFilter → AsCodeFilter (condition, group, or DSL)
- `toStoredFilter` - Converts AsCodeFilter → StoredFilter (preserving metadata for backwards compatibility)
  

### Key Features

1. **Complete Filter Type Coverage:**

   - **Condition filters**: phrase, phrases, range, exists
   - **Group filters**: combined filters with AND/OR relations and nested groups
   - **DSL filters**: custom queries, spatial filters (geo_shape), script queries, query_string

2. **Legacy Filter Migration:**

   - Handles legacy Kibana filter formats (top-level `range`, `match_phrase`, `exists`)
   - Migrates to modern query structure via `migrateFilter()`
   - Preserves backward compatibility for existing saved objects

3. **Metadata Preservation:**

   - Round-trip conversion maintains filter structure and critical metadata
   - Preserves critical properties: `isMultiIndex` (spatial filters), `field`/`params` (scripted filters)

4. **Date Range Support:**

   - Preserves date formats (`strict_date_optional_time_nanos`)
   - Supports time picker ranges (`now-15m` to `now`)
   - Handles single-value dates as equal gte/lte ranges

5. **Advanced Filter Handling:**

   - **Scripted filters**: Preserved as DSL with metadata for Painless scripts
   - **Spatial filters**: Maintains `isMultiIndex` and geo_shape query structure
   - **Negated phrases**: All filters can be negated with `negate: true`
   - **Complex bool queries**: Preserves queries with mixed must/should clauses as DSL


## Reviewers

Since this PR does not make any user-facing changes, one way to test the conversion is by testing the [Dashboard integration PR](https://github.com/elastic/kibana/pull/242775) which is a child of this branch specifically created for supporting AsCodeFilters in Dashboards. These conversion functions are used in that PR to return AsCodeFilters in the API response as well as converting AsCodeFilters to legacy Filter objects for backwards compatibility with the Unified Search bar in Dashboards.

nickpeihl committed:
Define schemas for simple filter interface

nickpeihl committed:
Add descriptions to schemas

nickpeihl committed:
Remove id and rename indexPattern to dataViewId

nickpeihl committed:
Merge remote-tracking branch 'upstream/main' into simple-filters-schema

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

nickpeihl committed:
Fix filter value schema arrays to be homogeneous

nickpeihl committed:
Fix casing inconsistency

nickpeihl committed:
Naming consistency

nickpeihl committed:
Stricter discriminated filter conditions

nickpeihl committed:
Document controlledBy field

nickpeihl committed:
Remove duplicate stored filter schema

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

nickpeihl committed:
Merge remote-tracking branch 'origin/simple-filters-schema' into simple-filters-transforms

nickpeihl committed:
Rename path

nickpeihl committed:
Merge remote-tracking branch 'refs/remotes/origin/simple-filters-schema' into simple-filters-schema

nickpeihl committed:
Add conversion functions from/to stored filters

nickpeihl committed:
Add es-query-constants package

This allows both `@kbn/es-query` and `@kbn/es-query-server` to use the same constants and avoids circular dependency issues

nickpeihl committed:
Merge remote-tracking branch 'origin/simple-filters-schema' into simple-filters-transforms

nickpeihl committed:
Rename stored filter schema to "storedFilterSchema" and "simpleFilterSchema" to "filterSchema"

nickpeihl committed:
Add constants package to avoid circular references

nickpeihl committed:
Add  isMultiIindex, filterType, and BWC legacy properties

kibanamachine committed:
Changes from node scripts/generate codeowners

nickpeihl committed:
Merge branch 'main' into simple-filters-schema

nickpeihl committed:
Rename AsCodeFilter types

nickpeihl committed:
Merge remote-tracking branch 'origin/simple-filters-schema' into simple-filters-schema

nickpeihl committed:
As "asCode" prefix to exported schemas

nickpeihl committed:
Don't export StoredFilter types

nickpeihl committed:
Use schema extend instead of spread

nickpeihl committed:
Remove unnecessary nested query object from DSL schema

nickpeihl committed:
Merge remote-tracking branch 'origin/simple-filters-schema' into simple-filters-transforms

nickpeihl committed:
Fix type names and fixtures

nickpeihl committed:
Update comment

nickpeihl committed:
Update src/platform/packages/shared/kbn-es-query-constants/package.json

Co-authored-by: Davis McPhee <davismcphee@hotmail.com>

nickpeihl committed:
Merge branch 'main' into simple-filters-schema

nickpeihl committed:
Merge remote-tracking branch 'origin/simple-filters-schema' into simple-filters-schema

nickpeihl committed:
Merge remote-tracking branch 'origin/simple-filters-schema' into simple-filters-transforms

nickpeihl committed:
Better round-trip stored filter conversion

nickpeihl committed:
Improve round-trip conversion

nickpeihl committed:
Merge remote-tracking branch 'upstream/main' into simple-filters-transforms

nickpeihl committed:
Clean up post merge

nickpeihl committed:
Clean up and bug fixes

nickpeihl committed:
Clean ups

nickpeihl committed:
Return early

nickpeihl committed:
More early returns

nickpeihl committed:
More cleanups

nickpeihl committed:
Test organization

nickpeihl committed:
Duplicate test cleanup

nickpeihl committed:
Merge remote-tracking branch 'upstream/main' into simple-filters-transforms

nickpeihl committed:
Throw early if stored filter is invalid

nickpeihl committed:
Clean up unused export

nickpeihl committed:
Refactor fromStoredFilter to use meta.type to determine conversion method

nickpeihl committed:
Update test

nickpeihl committed:
Simplify conversions

nickpeihl committed:
Minimize returning weak defaults

nickpeihl committed:
Remove `negate` property from condition filters.

Condition filter operators already determine negation. The top level negate on stored condition filters should not pass through to as code condition filters.

nickpeihl committed:
Remove pinned from AsCodeFilters

nickpeihl committed:
Support format in range type AsCodeFilters

nickpeihl committed:
Prefer consts

nickpeihl committed:
lint ts projects

nickpeihl committed:
OAS refinements

nickpeihl committed:
Remove unnecessary fixtures and tests used in debugging

nickpeihl committed:
Remove unused export

nickpeihl committed:
Test fixes

nickpeihl committed:
Remove unused types

nickpeihl committed:
Fix test types

nickpeihl committed:
Merge branch 'main' into simple-filters-transforms

nickpeihl committed:
Fix missing negation on range filters

nickpeihl committed:
AsCodeFilters should not include pinned stored filters

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/366111)
* Commit: 6338af3064a3017d0ed478de92adbd05e5d799cf

### Failed CI Steps
* [Store Cache for build](https://buildkite.com/elastic/kibana-pull-request/builds/366111#019ac21c-0744-4c09-b4a6-b7dee5424d55)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366111","number":366111,"commit":"6338af3064a3017d0ed478de92adbd05e5d799cf"}],"number":366111}
buildkite-pr-comment-->

nickpeihl committed:
Merge remote-tracking branch 'upstream' into simple-filters-transforms

nickpeihl committed:
Move AsCode filter validation schemas and types
from @kbn/es-query-server to new dedicated package
`@kbn/as-code-filters-schema`.

nickpeihl committed:
Move AsCode filter transform functions from `
@kbn/es-query` to new dedicated package
`@kbn/as-code-filters-transforms`.

nickpeihl commented on src/platform/packages/shared/as-code/filters-schema/kibana.jsonc:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "shared-common",
```
Ideally, this would be `shared-server`, but Lens will need to use this in their schemas which are currently defined in a `shared-common` package.

kibanamachine committed:
Changes from node scripts/generate codeowners

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

nickpeihl committed:
Move AS_CODE_FILTER_OPERATOR constant to its own
package

nickpeihl committed:
Merge branch 'simple-filters-transforms' of https://github.com/nickpeihl/kibana into simple-filters-transforms

kibanamachine committed:
Changes from node scripts/generate codeowners

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

nickpeihl committed:
Do not return undefined stored filter properties

nickpeihl committed:
Use snake_case for AsCodeFilter properties

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369194)
* Commit: cf78b9d24702476c266b2536fe878f8d19d5e515

### Failed CI Steps
* [FTR Configs #102](https://buildkite.com/elastic/kibana-pull-request/builds/369194#019aeadc-e931-46a7-b274-76e9d3eb274b)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/9785c55cec8345289dca6752a37b439c0e54db34) | [after](https://github.com/elastic/kibana/commit/cf78b9d24702476c266b2536fe878f8d19d5e515) | diff |
| --- | --- | --- | --- |
| `@kbn/as-code-filters-schema` | - | 4 | +4 |
| `@kbn/as-code-filters-transforms` | - | 10 | +10 |
| total |  |  | +14 |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/9785c55cec8345289dca6752a37b439c0e54db34) | [after](https://github.com/elastic/kibana/commit/cf78b9d24702476c266b2536fe878f8d19d5e515) | diff |
| --- | --- | --- | --- |
| `@kbn/as-code-filters-constants` | - | 1 | +1 |
| `@kbn/as-code-filters-schema` | - | 8 | +8 |
| `@kbn/as-code-filters-transforms` | - | 27 | +27 |
| `@kbn/es-query-server` | 10 | 8 | -2 |
| total |  |  | +34 |


</details>


### History
* :green_heart: [Build #368676](https://buildkite.com/elastic/kibana-pull-request/builds/368676) succeeded ae140e0d0323a5ee0947bb1e4b837b0e24ad5be9
* :broken_heart: [Build #366111](https://buildkite.com/elastic/kibana-pull-request/builds/366111) failed 6338af3064a3017d0ed478de92adbd05e5d799cf

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369194","number":369194,"commit":"cf78b9d24702476c266b2536fe878f8d19d5e515"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368676","number":368676,"commit":"ae140e0d0323a5ee0947bb1e4b837b0e24ad5be9"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366111","number":366111,"commit":"6338af3064a3017d0ed478de92adbd05e5d799cf"}],"number":369194}
buildkite-pr-comment-->

lukasolson commented on src/platform/packages/shared/as-code/filters-transforms/src/from_stored_filter.ts:
```diff
@@ -0,0 +1,513 @@
+    };
+  }
+
+  if (query.term) {
```
I'm not sure I am aware of how to get a filter that contains a `term` query... Is there something I'm missing?

nickpeihl replied:
Good point. I've removed this code. Any `term` queries will fallback to a DSL query.

lukasolson commented on src/platform/packages/shared/as-code/filters-transforms/src/from_stored_filter.ts:
```diff
@@ -0,0 +1,513 @@
+  }
+
+  if (query.match) {
+    const matchField = Object.keys(query.match)[0];
```
Same question here, I was under the impression we always generate phrase queries.

lukasolson reviewed (commented):
Just a handful of questions! Shouldn't be anything really blocking.

nickpeihl committed:
Merge remote-tracking branch 'upstream/main' into simple-filters-transforms

ThomThomson reviewed (commented):
Left a few comments, mostly focusing on the shapes of the new Filters themselves, haven't looked through the transformers in detail quite yet.

nickpeihl commented on src/platform/packages/shared/as-code/filters-transforms/src/from_stored_filter.ts:
```diff
@@ -0,0 +1,513 @@
+ * Helper to remove negate property from base properties
+ * For condition filters, negate is encoded in the operator (is_not, is_not_one_of, not_exists)
+ */
+function stripNegateProperty(
```
Based on a [suggestion by @ThomThomson below](https://github.com/elastic/kibana/pull/242043#discussion_r2688260211), I'm considering removing the `*NOT*` operators altogether and allowing `negate` on all conditions. This might be simpler and more consistent.

nickpeihl committed:
Remove negation operators in favor of negate property for all conditions

nickpeihl committed:
Simplify simple condition filter transform

nickpeihl committed:
Clean up negation handling (follow-up to 4afaeda)

lukasolson commented on src/platform/packages/shared/as-code/filters-transforms/src/to_stored_filter.ts:
```diff
@@ -0,0 +1,486 @@
+          ...(condition.value.gt !== undefined && { gt: condition.value.gt }),
+          ...(condition.value.lt !== undefined && { lt: condition.value.lt }),
+          ...(format && { format }),
+        },
```
```suggestion
        params: rangeQuery,
```

lukasolson commented on src/platform/packages/shared/as-code/filters-transforms/src/type_guards.ts:
```diff
@@ -0,0 +1,189 @@
+  exists?: unknown;
+  match_all?: unknown;
+  match?: unknown;
+}
```
Curious if this shape is implied by `migrateFilter`? I'm just wondering because I think it is very old (like before Kibana 6.0 or something wild) and I just want to make sure this is actually necessary.

lukasolson reviewed (commented):
Left some pieces of feedback below! This is a huge effort, thanks for being brave and willing to take this on!

nickpeihl committed:
Fix recursive type

nickpeihl committed:
Review feedback

nickpeihl committed:
Merge remote-tracking branch 'upstream' into simple-filters-transforms

nickpeihl committed:
Remove isLegacyFilter check.

Pass all filters through migrateFilter instead.

ThomThomson commented on src/platform/packages/shared/as-code/filters-transforms/README.md:
```diff
@@ -0,0 +1,230 @@
+
+**Smart Type Detection**:
+
+- Uses preserved `filterType` if available
```
How much drift does it introduce when `filterType` is dropped during the conversion? I'd really like to avoid releasing our new definition with deprecated fields. Seems overly cautious perhaps?

ThomThomson commented on src/platform/packages/shared/as-code/filters-transforms/README.md:
```diff
@@ -0,0 +1,230 @@
+
+**Group Optimization**:
+
+- OR group with same-field IS conditions → `phrases` filter (compact representation)
```
Nice detail!

ThomThomson reviewed (commented):
I have a few questions about the deprecated items in the schema. If at all possible I'd very much like to avoid our _new_ schema for as-code filters having any deprecated fields or any fields that exist solely to retain BWC information.

nickpeihl committed:
Remove unnecessary `value` property from as code schema

nickpeihl committed:
Use an actual discriminated union for as code filters by adding a `type` property

nickpeihl committed:
Add spatial filter type similar to dsl but preserves SPATIAL type in stored filter meta

nickpeihl committed:
Remove `filter_type` from as code filters using logic to preserve stored `meta.type`

nickpeihl committed:
Remove `key` field from as code filter schema

nickpeihl committed:
Remove "deprecated" from optional DSL filter properties

nickpeihl committed:
Use constants for filter types

nickpeihl committed:
Remove unused type guard

nickpeihl committed:
Remvoe unnecessary detection of filter type from custom DSL filters

nickpeihl committed:
Merge branch 'main' into simple-filters-transforms

ThomThomson reviewed (approved):
Excellent work and response to / incorporation of all the previous feedack. Looked through the code and schemas again, and tested locally by running [this PR](https://github.com/elastic/kibana/pull/242775) and everything is looking and feeling great. Seeing `filters` come back in the network tab with a **clean schema** is huge, and I'm starting to feel like this is a step towards fixing one of the biggest irks of Kibana in general. Very nice set of changes, LGTM!

nickpeihl committed:
Merge branch 'main' into simple-filters-transforms

lukasolson reviewed (commented):
Left a few minor comments below, overall this is looking really good!

nickpeihl committed:
Use `operator` instead of `type` for condition filters

nickpeihl committed:
Remove unreachable code

nickpeihl committed:
Add discriminator to conditionSchema

nickpeihl committed:
Function name clarification

nickpeihl committed:
Merge branch 'main' into simple-filters-transforms

lukasolson reviewed (approved):
LGTM! Thanks for pushing through these last few changes.
